erpnext/erpnext_integrations/doctype/leaflink_settings/leaflink_settings.py

```diff
@@ -0,0 +1,182 @@
+	}
+	sales_order = args.get("data")
+
+	#ensure leaflink is enabled before proceeding further.
+	if leaflink_settings.get("enable_leaflink") != 1:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Leaflink is not enabled."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+	#validate incoming keys
+	key = bytes(leaflink_settings.get("incoming_api_authentication_key"), 'utf-8')
+	incoming_signature = frappe.local.request.headers['LL-Signature'].encode("utf-8")
+	request_body = frappe.local.request.data
+	computed_signature = base64.b64encode(hmac.new(key, request_body, digestmod=hashlib.sha256).digest())
+	success = (computed_signature == incoming_signature)
+	if not success:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Signatures do not match. Please check the Incoming Key in LeafLink Settings."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+
+	#exit if the action is not to create a new sales order
+	if args.get("action") != "create":
+		return
+
+	#extract customer and license data
+	customer = sales_order.get("customer")
+	customer_name = customer.get('name')
+	license_no = customer.get('license_number')
+
+	#retrieve customer_name if license_no is in record
+	if not frappe.db.exists("Customer", customer_name):
+
+		license_no = customer.get('license_number')
+		if frappe.db.exists("Compliance Info", license_no):
+
+			#find customer using license info in case name does not match
+			license_parent = frappe.get_all("Compliance License Detail",
+				filters={"license":license_no},
+				fields=['parent'])
+			if license_parent:
+				customer_name = license_parent[0].get("parent")
+		else:
+			#update integration request with message telling user that the customer needs to be created before proceeding further
+			integration_request.update({
+				"status": "Failed",
+				"error": "Customer does not exist in Bloomstack. Please create the customer and try again."
+			})
+			integration_request.save(ignore_permissions=True)
+			return
+
+
+	#create a sales order document with an autoset expected delivery date
+	expected_delivery_days = int(frappe.get_value("Selling Settings", None, "expected_delivery_days")) or 7
+	sales_order_doc = frappe.get_doc({
+		"doctype": "Sales Order",
+		"transaction_date": datetime.datetime.now(),
+		"customer": customer_name,
+		"license": license_no,
+		"delivery_date": add_days(datetime.datetime.now(), expected_delivery_days),
+		"order_type": "Sales"
+	})
+
+	#visit all items in order from leaflink, match sku to item code and add to sales order
+	for item in sales_order.get("orderedproduct_set"):
+
+		#fetching important information from leaflink order
+		item_code = item.get("product").get("sku")
+		rate = item.get("sale_price")
+		quantity = item.get("quantity")
+
+		#appending items to sales order document
+		sales_order_doc.append("items",{
+			"item_code": item_code,
+			"rate": rate,
+			"qty": quantity,
```

Comment:
```suggestion
			"item_code": item.get("product").get("sku"),
			"rate": item.get("sales_price"),
			"qty": item.get("quantity"),
```

Reply:
we can do the .get() directly here as its used only once